background/services/chain/taho-rpc-provider.ts

```diff
@@ -0,0 +1,316 @@
+import { JsonRpcProvider, Networkish } from "@ethersproject/providers"
+import { deepCopy } from "@ethersproject/properties"
+import { ConnectionInfo, fetchJson } from "@ethersproject/web"
+import logger from "../../lib/logger"
+import { wait } from "../../lib/utils"
+
+type RPCPayload = {
+  method: string
+  params: unknown[]
+  id: number
+  jsonrpc: "2.0"
+}
+
+type RPCResponse = RPCResponseResult | RPCResponseError
+
+type RPCResponseError = {
+  id: number
+  error: {
+    code: number
+    message: string
+    data?: unknown
+  }
+  jsonrpc: "2.0"
+}
+
+type RPCResponseResult = {
+  id: number
+  result: unknown
+  jsonrpc: "2.0"
+}
+
+type RPCPendingRequest = {
+  resolve: (value: unknown) => void
+  reject: (reason: unknown) => void
+  payload: RPCPayload
+}
+
+type RPCOptions = {
+  /**
+   * Maximum number of requests to be batched
+   * @default 10 requests
+   */
+  maxBatchLength?: number
+  /**
+   * Maximum length in bytes of the batch
+   * @default 1Mb
+   */
+  maxBatchSize?: number
+  /**
+   * How long to wait to aggregate requests
+   * @default 100ms
+   */
+  batchStallTime?: number
+  /**
+   * How long to wait between each payload sent
+   */
+  delayBetweenRequests?: number
+}
+
+const defaultOptions = {
+  maxBatchLength: 10, // Seems to work fine for public rpcs
+  // eslint-disable-next-line no-bitwise
+  maxBatchSize: 1 << 20, // 1Mb
```

Comment:
```suggestion
  1_048_576, // 1Mb
```

Why not just write this out?